optimade/validator/validator.py

```diff
@@ -712,7 +713,10 @@ def _construct_single_property_filters(
                 )
 
             response = response.json()
-            num_data_returned[operator] = response["meta"]["data_returned"]
+            if not response["meta"]["more_data_available"]:
+                num_data_returned[operator] = len(response["data"])
+            else:
+                num_data_returned[operator] = response["meta"].get("data_returned")
```

Comment:
I am unsure what is expected to be present in the response at this point, so I am just suggesting this to try and make extra sure it's all OK.
Furthermore, I think "positive" if-conditions are more readable.
There's some fallbacks needed for these keys as well.
```suggestion
            if response.get("meta", {}).get("more_data_available", False):
                num_data_returned[operator] = response.get("meta", {}).get("data_returned", data_returned)
            else:
                num_data_returned[operator] = len(response["data"])
```

Reply:
`meta` should ("MUST") be there, as should `more_data_available`, but `data_returned` should then be checked against `None` later on and not used unless `len(response["data"])` matches the value of data_returned (i.e. it is less than the page limit), so don't think I want to change this.

Reply:
Since `more_data_available` is not MUST, it should probably not be expected to be present. But yeah, `meta` is a MUST and as such can be expected to be present.
Do we check that it actually _is_ present and avoid doing all this if that's the case?

Concerning the changes to the if/else, the only thing that's changed is the switch from checking a negative to testing a positive.

Reply:
`more_data_available` is a MUST?

Reply:
The negative testing is pretty subjective, I think it's clearer to have the fallback (checking `data_returned`) as the second part of the clause, though I would agree with you if the condition was more complicated than checking the value of a bit...

Reply:
> `more_data_available` is a MUST?

Not as I read it here: https://github.com/Materials-Consortia/OPTIMADE/blob/develop/optimade.rst#json-response-schema-common-fields
It seems `meta` is a MUST, but there are no MUST keys in it, only SHOULD and MAY.

Reply:
> The negative testing is pretty subjective, I think it's clearer to have the fallback (checking `data_returned`) as the second part of the clause, though I would agree with you if the condition was more complicated than checking the value of a bit...

My main motivation being readability for double negatives. But yes, it's very subjective, which is why I didn't originally review this as a "Changes requested", but rather as a "Comment" :)

Reply:
That's not my reading of this?

> meta: a JSON API meta member that contains JSON API meta objects of non-standard meta-information. It MUST be a dictionary with these fields:
>
>    more_data_available: false if the response contains all data for the request (e.g., a request issued to a single entry endpoint, or a filter query at the last page of a paginated response) and true if the response is incomplete in the sense that multiple objects match the request, and not all of them have been included in the response (e.g., a query with multiple pages that is not at the last page).

Reply:
Be careful of the distinction between `data_available` (which is optional, c.f. the other commit in this PR) and `more_data_available` which is a MUST

Reply:
You're right! I read the line above `meta` and took it to mean something different, but it was _only_ referring to the top-level fields, not the contents of those fields.
Cheers. It states it quite clearly in the first line you cited above.